### PR TITLE
DEVX-87 Migrate ProductTypeReferenceResolutionUtils

### PR DESCRIPTION
**Pull request by lojzatran**
#### Summary

JIRA: https://commercetools.atlassian.net/browse/DEVX-87

### CODE REVIEW

**Review by salander85 — CHANGES_REQUESTED** [review]

File: src/main/java/com/commercetools/sync/sdk2/producttypes/utils/ProductTypeReferenceResolutionUtils.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+              return toAttributeDefinitionDraftBuilder(attributeDefinition)
+                  .type(referenceReplacedType)
+                  .build();
+            })
+        .filter(Objects::nonNull)
+        .collect(toList());
+  }
+
+  @Nonnull
+  private static AttributeType replaceIdWithKeyForProductTypeReference(
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/sdk2/producttypes/utils/ProductTypeReferenceResolutionUtils.java:101]: You are right. However, it was also not in the current version with SDK v1: https://github.com/commercetools/commercetools-sync-java/blob/feec85440fc695c26023a32374359eb269ca914b/src/main/java/com/commercetools/sync/producttypes/utils/ProductTypeReferenceResolutionUtils.java#L99.

[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/sdk2/producttypes/utils/ProductTypeReferenceResolutionUtils.java:101]: I ran an integration test where an attribute reference is created and the product type looks like this:
```json
    {
      "name": "name_1",
      "description": "description_1",
      "classifier": "Complex",
      "attributes": [
        {
          "name": "referenceTypeAttr",
          "label": {
            "en": "referenceTypeAttr"
          },
          "isRequired": false,
          "type": {
            "name": "set",
            "elementType": {
              "name": "reference",
              "referenceTypeId": "category"
            }
          },
          "attributeConstraint": "None",
          "isSearchable": false,
          "inputHint": "SingleLine",
          "displayGroup": "Other"
        }
      ],
      "key": "key_1"
    }
```
There is nothing to resolve in `AttributeReferenceType` so the code is correct. The real reference is not in the product type definition, but in the product when the attr is used.

File: src/main/java/com/commercetools/sync/sdk2/producttypes/helpers/ResourceToDraftConverters.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        .name(productType.getName())
+        .attributes(
+            productType.getAttributes().stream()
+                .map(
+                    attributeDefinition ->
+                        toAttributeDefinitionDraftBuilder(attributeDefinition).build())
+                .collect(toList()));
+  }
+
+  public static AttributeDefinitionDraft toAttributeDefinitionDraft(
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/sdk2/producttypes/helpers/ResourceToDraftConverters.java]: https://github.com/commercetools/commercetools-sync-java/pull/980/commits/bc79e7314111804b4ca0118f974c67b92936b87e

File: src/main/java/com/commercetools/sync/sdk2/producttypes/utils/ProductTypeReferenceResolutionUtils.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+   *        <td>productType references (in case it has NestedType or set of NestedType)</td>
+   *        <td>{@link java.util.Set}&lt;{@link com.commercetools.api.models.product_type.ProductTypeReference}&gt;</td>
+   *        <td>{@link java.util.Set}&lt;{@link com.commercetools.api.models.product_type.ProductTypeReference}&gt; (with key replaced with id field)</td>
+   *     </tr>
+   *   </tbody>
+   * </table>
+   *
+   * <p><b>Note:</b>If some references are not expanded for an attributeDefinition of a productType,
+   * the method will throw a {@link
+   * com.commercetools.sync.commons.exceptions.ReferenceReplacementException} containing the root
```
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/main/java/com/commercetools/sync/sdk2/producttypes/utils/ProductTypeReferenceResolutionUtils.java]: https://github.com/commercetools/commercetools-sync-java/pull/980/commits/b36a6e0393fec54754b7213d98e4d1a8e13fd1ac

**Review by salander85 — APPROVED** [review]

**Pull request merged**